---
title: "double/single rate ratio"
author: "Arlin Stoltzfus"
date: "2023-09-12"
output: html_document
---

# Double/single rate ratio estimates used by Klug, et al

This code is a supplement to Klug A, Avella M, Cano AV, Payne JL, Stoltzfus A, McCandlish, DM. in prep. Less frequent but more numerous: how rare mutational changes contribute to adaptation in finite mutational landscapes.

Please direct any feedback to Arlin Stoltzfus (arlin\@umd.edu)

## Overview

Below are estimates of the double/single mutation rate ratio (in humans) from the ratio of counts of double (tandem) and single base substitutions among mutations called from genomic data. The results are summarized in a single table with the following information: 

* name: an informal identifier composed from first author and year
* ratio: the double/single ratio rounded to 5 digits of precision
* type: germline, somatic, or cancerous
* scope: typically whole-genome, but sometimes only coding regions
* bibcode: a bibliographic code used internally 
* comment: an explanation, or a quoted statement with the numbers ("quote{<statement>}")

The published sources listed in the Bibliography below are unambiguously identified by the name or bibcode. 

This code serves to document precisely how each estimate was derived because, with the exception of Chen, et al, all of the values were derived from calculations, rather than being stated explicitly by the authors.  In some cases, the calculation is as simple as taking the reported number of doubles and dividing by the reported number of singles. In other cases, we need to identify and sum events from supplementary data files. 

Often when estimating rates or frequencies from differences, there is a correction for multiple hits. That is not required here because the sampling frequency is far too low for this to matter. We are only aiming for 1 or 2 digits of precision. 

## Germline 

Meta-analysis by Chen, et al

```{r}

alphas <- data.frame(alpha = 0.004, type = "germline", scope = "whole genome", bibcode = "Chen-2014-6836", comment = "meta-analysis")
rownames(alphas) <- "Chen2014"
```

## Cancer cells

```{r cancer}
name <- "Alexandrov2020"
bibcode <- "Alexandrov-2020-6852"
type <- "cancer (mixed types)"
scope <- "whole genome"
comment <- "quote{The 23,829 samples—which include most types of cancer, and comprise the 2,780 PCAWG whole genomes2, 1,865 additional whole genomes and 19,184 exomes—yielded 79,793,266 somatic SBSs, 814,191 doublet-base substitutions (DBSs) and 4,122,233 small indels that were analysed for mutational signatures}"

dbs <- 814191
sbs <- 79793266
ratio <- round(dbs/sbs, digits = 5)

# add results to table
alphas <- rbind(alphas, c(ratio, type, scope, bibcode, comment))
rownames(alphas)[dim(alphas)[1]] <- name

name <- "Degasperi2022"
bibcode <- "Degasperi-2022-6853"
type <- "cancer (mixed types)"
scope <- "whole genome"
comment <- "quote{For this evaluation, the final dataset included 298,694,545 substitutions, 2,675,617 double substitutions, 154,675,475 indels, and 1,958,105 rearrangements}"

dbs <- 2675617
sbs <- 298694545
ratio <- round(dbs/sbs, digits = 5)

# add results to table
alphas <- rbind(alphas, c(ratio, type, scope, bibcode, comment))
rownames(alphas)[dim(alphas)[1]] <- name

name <- "Everall2023"
bibcode <- "Everall-2023-6855"
type <- "cancer (mixed types)"
scope <- "whole genome"
comment <- "quote{The final dataset comprised 285,233,025 somatic single base substitution (SBS) mutations, 2,272,616 doublet base substitution (DBS) mutations, 82,161,895 small insertions and deletions (ID), 1,285,875 copy number (CN) alterations and 811,962 structural variants (SVs).}"

dbs <- 2272616
sbs <- 285233025
ratio <- round(dbs/sbs, digits = 5)

# add results to table
alphas <- rbind(alphas, c(ratio, type, scope, bibcode, comment))
rownames(alphas)[dim(alphas)[1]] <- name

```

# Non-cancerous somatic cells

## Lee-Six whole genome

Here I am just summing over the tables for singlets and doublets, then dividing to get the ratio. 

```{r lee-six}
name <- "Lee-Six2019-whole"
bibcode <- "Lee-Six-2019-6857"
type <- "somatic (colorectal)"
scope <- "whole genome"
comment <- "whole genome called mutations from colorectal epithelial cells"

# downloaded the infiles from GitBub https://github.com/HLee-Six/colon_microbiopsies
datadir <- "Lee-Six-data-analysis"
sbsfile <- "sbs_category_counts.txt"
dbsfile <- "dbs_category_counts.txt"

sbsdat <- read.csv(paste(datadir, sbsfile, sep = "/"), stringsAsFactors = F, sep = "\t")
sbscnt <- sum(sbsdat)

dbsdat <- read.csv(paste(datadir, dbsfile, sep = "/"), stringsAsFactors = F, sep = "\t")
dbscnt <- sum(dbsdat)

dbscnt
sbscnt
ratio <- round(dbscnt/sbscnt, digits = 5)

# add results to table
alphas <- rbind(alphas, c(ratio, type, scope, bibcode, comment))
rownames(alphas)[dim(alphas)[1]] <- name

```

## Mitchell, et al blood cells, coding only 

I did not understand how they are coding up mutations, so I asked Iñigo Martincorena and he explained how to process the files and provided some R code, which is integrated below. 

```{r mitchell}
name <- "Michell2022-coding"
bibcode <- "Mitchell-2022-6860"
type <- "somatic (blood)"
scope <- "coding regions"
comment <- "coding region mutations from blood cells"
supfile <- "41586_2022_4786_MOESM11_ESM.csv"
datadir <- "Mitchell-2022-data-analysis"

# Load mutations
m = read.table(paste(datadir, supfile, sep="/"), header=1, sep=",", stringsAsFactors = F) 

# m2 = read.table("../Lee-Six-data-analysis/41586_2019_1672_MOESM6_ESM.csv", header=1, sep=",", stringsAsFactors = F) 
# m = m2

# Sort mutations by sampleID, chr, pos 
# and remove duplicates introduced by the annotation with dndscv
m = unique(m[order(m$sampleID,m$chr,m$pos),1:5]) 

# Create a new index d from pos by subtracting a number that increases by 1 in each row,
# so that values of d will be the same when successive pos values are in successive rows, 
# e.g., [ 29, 30, 31 ] minus [ 3, 4, 5 ] gives [ 26, 26, 26 ]
d = m$pos-(1:nrow(m))
runs = rle(d)

# the numbers of apparent singles and doubles
table(runs$lengths)

# initialize an index for rows to be removed from m
rmpos = rep(0,nrow(m))
# create an index for the start of each run
runstarts = cumsum(runs$length)-runs$length+1
for (h in 1:length(runs$length)) {
  if (runs$length[h]>1) { # Adjacent mutations
    mutcluster = runstarts[h]:(runstarts[h]+runs$lengths[h]-1)
    # Removing all the affected rows except the first one (which we will edit to capture the complex event)
    rmpos[mutcluster[-1]] = 1 
    m[mutcluster[1],"ref"] = paste(m[mutcluster,"ref"],collapse="")
    m[mutcluster[1],"mut"] = paste(m[mutcluster,"mut"],collapse="")
  }
}
# remove extra rows from m
m = m[!rmpos,]
# for convenience, show just the doubles
dbstable = m[nchar(m$ref) == 2 & nchar(m$mut) == 2,]
print(dbstable)

```
Now compute Mitchell results and add to file

```{r mitchell-out}
dbs <- sum(nchar(m$ref) == 2 & nchar(m$mut) == 2)
sbs <- sum(nchar(m$ref) == 1 & nchar(m$mut) == 1)
ratio <- round(dbs/sbs, digits = 5)

# add results to table
alphas <- rbind(alphas, c(ratio, type, scope, bibcode, comment))
rownames(alphas)[dim(alphas)[1]] <- name
alphas[, 1:4]
```
## Coding mutations from Lee-Six

The corresponding supplement from Lee-Six, et al. does not have the doubles in it, for some reason.  Iñigo Martincorena was again helpful in explaining where to get the ratio, which simply involves summing over some columns. 

```{r lee-six-again}
name <- "Lee-Six2019-coding"
bibcode <- "Lee-Six-2019-6857"
type <- "somatic (colorectal)"
scope <- "coding regions"
comment <- "coding region called mutations from colorectal epithelial cells"
supfile <- "41586_2019_1672_MOESM5_ESM.txt"
datadir <- "Lee-Six-data-analysis"
# Load mutations
mydat = read.table(paste(datadir, supfile, sep="/"), header=1, sep="\t", stringsAsFactors = F) 

dbs <- sum(mydat$dbstotal)
sbs <- sum(mydat$sbstotal)
ratio <- round(dbs/sbs, digits = 5)

# add results to table
alphas <- rbind(alphas, c(ratio, type, scope, bibcode, comment))
rownames(alphas)[dim(alphas)[1]] <- name
# alphas[, 1:4]
```

```{r output}
write.csv(alphas, file = "alpha-estimates.csv", row.names = T)
```


## Bibliography

Alexandrov LB, Nik-Zainal S, Wedge DC, Aparicio SA, Behjati S, Biankin AV, Bignell GR, Bolli N, Borg A, Borresen-Dale AL, et al. 2013. Signatures of mutational processes in human cancer. Nature 500:415-421.

Chen J-M, Cooper DN, Férec C. 2014. A New and More Accurate Estimate of the Rate of Concurrent Tandem-Base Substitution Mutations in the Human Germline: ∼0.4 % of the Single-Nucleotide Substitution Mutation Rate. Human Mutation 35:392-394.

Degasperi A, Zou X, Amarante TD, Martinez-Martinez A, Koh GCC, Dias JML, Heskin L, Chmelova L, Rinaldi G, Wang VYW, et al. 2022. Substitution mutational signatures in whole-genome-sequenced cancers in the UK population. Science 376.

Everall A, Tapinos A, Hawari A, Cornish A, Sud A, Chubb D, Kinnersley B, Frangou A, Barquin M, Jung J, et al. 2023. Comprehensive repertoire of the chromosomal alteration and mutational signatures across 16 cancer types from 10,983 cancer patients. medRxiv 2023.06.07.23290970w.

Klug A, Avella M, Cano AV, Payne JL, Stoltzfus A, McCandlish, DM. in prep. Less frequent but more numerous: how rare mutational changes contribute to adaptation in finite mutational landscapes.

Lee-Six H, Olafsson S, Ellis P, Osborne RJ, Sanders MA, Moore L, Georgakopoulos N, Torrente F, Noorani A, Goddard M, et al. 2019. The landscape of somatic mutation in normal colorectal epithelial cells. Nature 574:532-537.

Mitchell E, Spencer Chapman M, Williams N, Dawson KJ, Mende N, Calderbank EF, Jung H, Mitchell T, Coorens THH, Spencer DH, et al. 2022. Clonal dynamics of haematopoiesis across the human lifespan. Nature 606:343-350.

# Rates per cell per generation, or per year

Ignore this. Not the double/single ratio. This is for something else

```{r absolute-rates}

Blokzijl2016 <- "quote{Our results show that mutations accumulate steadily over time in all of the assessed tissue types, at a rate of approximately 40 novel mutations per year, despite the large variation in cancer incidence among these tissues1.}"

mutscount <- 40
genomesize <- 6.4E9
mutscount/genomesize

# vs 1 female to 3 male mutations per year in germline cells
manders2021 <- "quote{The amount of mutations that accumulate in different tissues ranged from 9 substitutions per year in bile ductular cells to 56 substitutions per year in appendiceal crypts (Moore et al., 2021). Mutation rates in other tissues fell within this range, showing that while there are differences between tissues, they all fall within a single order of magnitude (Figure 1). Female and male germ cells acquire only 0.74 and 2.7 mutations per year, showing that the mutation rate in somatic cells is much higher than in germline cells (Rahbari et al., 2016; Jónsson et al., 2017; Moore et al., 2021).}"
```
